### PR TITLE
test: pin Llama3 TurboAttention GQA layout (PRPUNDIT-20)

### PR DESCRIPTION
**Pull request by jiagaoxiang**
Same-repo refile of #1090 so Primus-CI-TAS runs against `AMD-AGI/Primus` (fork PRs do not receive Docker Hub credentials, so `build-docker` / torch unit tests never ran).

This PR was created by dougljia via Test Gap Resolver.

## Summary
- Adds `tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py`.
- Stub `inner_attention` as an `nn.Module` and assert unrepeated KV heads in BSHD (no `repeat_kv` / BHSD transpose), `positions` forwarded to RoPE, and that output still goes through `wo`.

Closes test gap **PRPUNDIT-20**.

## Test plan
- [x] `PYTHONPATH=. pytest tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py` (5 passed, CPU, needs torchtitan)
- [x] `black --check` / `isort --profile black` on the new file

### CODE REVIEW

File: tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        x = torch.randn(bs, seqlen, llama3_args.dim)
+        freqs_cis = torch.ones(seqlen, head_dim // 2, dtype=torch.complex64)
+
+        out = attn.forward(x, freqs_cis, attention_masks=None)
+
+        xq = attn.wq(x).view(bs, seqlen, -1, head_dim)
+        xk = attn.wk(x).view(bs, seqlen, -1, head_dim)
+        # Build `expected` via the same `apply_rotary_emb` call `Attention.forward`
+        # makes, rather than assuming freqs_cis=1 makes RoPE an exact no-op (an
+        # implementation detail of `apply_rotary_emb` this test shouldn't rely on).
```
Comment (jiagaoxiang) [inline tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py:162]: **TestIntel PR Steward:** Fixed in 9ff8021 — `expected` is now built by calling the real `apply_rotary_emb(xq, xk, freqs_cis=freqs_cis, positions=None)` (same call `Attention.forward` makes) instead of assuming `freqs_cis=1` is an exact RoPE no-op. Thanks for catching this.

File: tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+        # Should not raise even though `positions` is omitted.
+        out = attn.forward(x, freqs_cis, attention_masks=None)
+        assert out.shape == (bs, seqlen, llama3_args.dim)
+
+    def test_forward_uses_wo_projection(self, llama3_args):
+        # Sanity check that the mirror still routes through the base
+        # `Attention.wo` output projection rather than returning the raw
+        # inner_attention output.
+        import primus.backends.torchtitan.models.llama3.model.model as llama3_mirror
```
Comment (jiagaoxiang) [inline tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py:146]: **TestIntel PR Steward:** Partially addressed as a side effect of the fix above — `test_forward_uses_wo_projection` now also uses the `import ... as llama3_mirror` form since it needs `llama3_mirror.apply_rotary_emb`. I'm leaving the other three `from ... import Attention` occurrences as-is: this is a non-blocking style/lint nit with no behavioral effect, and converting the remaining tests is out of scope for this test-gap PR. Happy to do it if a human reviewer asks.

**Review by jiagaoxiang — COMMENTED** [review]
**TestIntel PR Steward:** Addressed the Copilot review comment (brittle RoPE no-op assumption in `test_forward_uses_wo_projection`) by pushing 9ff8021, which builds `expected` via the real `apply_rotary_emb` call instead of assuming `freqs_cis=1` is a no-op. Replied inline on both open threads — see below for the import-consistency nit, which I'm leaving mostly as-is (non-blocking style, out of scope for a test-only PR) except for the one occurrence the fix itself needed to touch.

CI: the current same-repo run (Primus-CI-TAS) has `dependency-review`, `code-lint`, and `build-docker` green; `run-unittest-jax`/`run-unittest-torch` were still queued as of this check — no actual CI failure on this SHA once the same-repo refile is running (the earlier `build-docker` failure was on the superseded fork PR, which never got Docker Hub credentials). Will keep watching.

File: tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        attn.forward(x, freqs_cis, attention_masks=None, positions=positions)
+
+        # Value-equivalence (not object identity): `Attention.forward` is only
+        # contracted to forward the same position values to RoPE, not the same
+        # tensor object (a harmless `.to(device)`/clone shouldn't break this).
+        assert captured["positions"] is not None
+        assert torch.equal(captured["positions"], positions)
+
+    def test_forward_defaults_positions_to_none(self, llama3_args):
+        from primus.backends.torchtitan.models.llama3.model.model import Attention
```
Comment (jiagaoxiang) [inline tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py:128]: **TestIntel PR Steward:** Same call as before on the other two threads: this is a repeat of the mixed `import X` / `from X import Y` style nit (now surfaced against `test_forward_defaults_positions_to_none`). It's non-blocking lint-only, no behavioral effect, and converting the remaining call sites is out of scope for this test-gap PR. Leaving as-is; happy to normalize all of them together if a human reviewer wants it.

File: tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        freqs_cis = torch.ones(seqlen, head_dim // 2, dtype=torch.complex64)
+        positions = torch.arange(seqlen).unsqueeze(0)
+
+        attn.forward(x, freqs_cis, attention_masks=None, positions=positions)
+
+        # Value-equivalence (not object identity): `Attention.forward` is only
+        # contracted to forward the same position values to RoPE, not the same
+        # tensor object (a harmless `.to(device)`/clone shouldn't break this).
+        assert captured["positions"] is not None
+        assert torch.equal(captured["positions"], positions)
```
Comment (jiagaoxiang) [inline tests/unit_tests/backends/torchtitan/test_llama3_turbo_attention.py:125]: **TestIntel PR Steward:** Fixed in db8cc3f — this now asserts `torch.equal(captured["positions"], positions)` instead of `is`, so a harmless `.to(device)`/clone in `Attention.forward` wouldn't spuriously fail the test. The intent (positions are forwarded, unchanged) is preserved. Thanks for flagging.

**Review by jiagaoxiang — COMMENTED** [review]
**TestIntel PR Steward:** Follow-up on the second Copilot pass (review #5115710135) — replied inline on both new threads. Summary:
- Fixed: object-identity assertion on `positions` in `test_forward_passes_positions_to_rotary_emb` → value-equivalence via `torch.equal` (commit db8cc3f).
- Declined: the recurring mixed-import-style nit against `test_forward_defaults_positions_to_none` — same as the earlier decision, non-blocking lint-only, out of scope for this test-gap PR.

CI on this SHA: `dependency-review`, `code-lint`, `build-docker` were green as of the last push; `run-unittest-jax`/`run-unittest-torch` were still queued. Will keep watching for CI completion and any further review activity, and merge once a human approval lands alongside green required checks.

**Review by Xiaoming-AMD — APPROVED** [review]
